### PR TITLE
chore: fix drift seen in xgboost regressor

### PR DESCRIPTION
**Pull request by jfrery**
closes https://github.com/zama-ai/concrete-ml-internal/issues/4076
closes https://github.com/zama-ai/concrete-ml-internal/issues/2604
closes https://github.com/zama-ai/concrete-ml-internal/issues/4079

### CODE REVIEW

**Review by RomanBredehoft — APPROVED** [review]
yes looks good to me, thanks a lot for debugging !

**Review by RomanBredehoft — DISMISSED** [review]
not sure how many examples we have with the XGB regressor but could be worth to update some docs / results in or after this PR

File: src/concrete/ml/sklearn/xgb.py [hunk 1]
```diff
@@ -331,6 +331,14 @@ def fit(self, X, y, *args, **kwargs) -> Any:
         super().fit(X, y, *args, **kwargs)
         return self
 
+    def post_processing(self, y_preds: numpy.ndarray) -> numpy.ndarray:
+        y_preds = super().post_processing(y_preds)
+
+        # Hummingbird Gemm for XGBoostRegressor adds a + 0.5 at the end of the graph.
+        # We need to add it back here since the graph is cut before this add node.
+        y_preds += 0.5
```
Comment (kcelia) [inline src/concrete/ml/sklearn/xgb.py:339]: That's what I was looking for in https://github.com/zama-ai/concrete-ml/pull/355/files lol

thanks

**Review by kcelia — DISMISSED** [review]

File: src/concrete/ml/sklearn/xgb.py (hunk 1 above)
Comment (andrei-stoian-zama) [inline src/concrete/ml/sklearn/xgb.py:339]: are you sure it's always 0.5 ? Can you maybe add an assert in the onnx parser ?

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]

File: src/concrete/ml/sklearn/xgb.py (hunk 1 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/xgb.py:339]: It's a constant being added called `base_prediction`. We can add an assert yes.

Comment (jfrery) [reply src/concrete/ml/sklearn/xgb.py:339]: Done.

File: src/concrete/ml/sklearn/tree_to_numpy.py [hunk 2]
```diff
@@ -85,6 +85,36 @@ def workaround_squeeze_node_xgboost(onnx_model: onnx.ModelProto):
     onnx_model.graph.node[target_node_id_list[0]].input.insert(1, axes_input_name)
 
 
+def assert_add_node_and_constant_in_xgboost_regressor_graph(onnx_model: onnx.ModelProto) -> bool:
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/tree_to_numpy.py:88]: remove "-> bool", this function does not return anything !

File: src/concrete/ml/sklearn/tree_to_numpy.py [hunk 3]
```diff
@@ -183,6 +213,12 @@ def tree_onnx_graph_preprocessing(
         on_error_msg=f"{len(onnx_model.graph.output)} != 2",
     )
 
+    # Check that a XGBoostRegressor onnx graph has the + 0.5 add node.
+    if framework == "xgboost":
+        # Make sure it's a regression and so a single onnx output.
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/tree_to_numpy.py:218]: "and so a single onnx output." weird sentence ?

File: src/concrete/ml/sklearn/tree_to_numpy.py (hunk 2 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/tree_to_numpy.py:88]: Indeed thanks

File: src/concrete/ml/sklearn/tree_to_numpy.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/tree_to_numpy.py:218]: Hmm sounds good to me. What do you suggest?

File: src/concrete/ml/sklearn/tree_to_numpy.py [hunk 4]
```diff
@@ -85,6 +85,36 @@ def workaround_squeeze_node_xgboost(onnx_model: onnx.ModelProto):
...
+    initializer_present = any(
+        [init for init in onnx_model.graph.initializer if init.name == constant_add_name]
+    )
+
+    # Iterate over all nodes in the model's graph
+    for node in onnx_model.graph.node:
+        # Check if the node is an "Add" node and has the
+        # specified initializer as one of its inputs
+        if node.op_type == "Add" and constant_add_name in node.input:
+            is_expected_add_node_present = True
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/sklearn/tree_to_numpy.py:116]: done

File: src/concrete/ml/sklearn/tree_to_numpy.py (hunk 2 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/tree_to_numpy.py]: done

File: src/concrete/ml/sklearn/tree_to_numpy.py (hunk 3 above)
Comment (RomanBredehoft) [reply src/concrete/ml/sklearn/tree_to_numpy.py]: ah ok got it, I would maybe rephrase it it still , like : `Make sure it's a regression model (by checking it has a single output, as mentioned above)` or something like it

Comment (jfrery) [reply src/concrete/ml/sklearn/tree_to_numpy.py]: done

**Review by RomanBredehoft — APPROVED** [review]
thanks a lot !

**Review by kcelia — APPROVED** [review]

**Review by andrei-stoian-zama — APPROVED** [review]
Thanks!
